Handle rangepos of by-name right-associative applications properly.

hrhino opened the pull request:
And, as a bonus, those with infix type args.

The by-name case is a simple tweak to the SIP-34 translation: the pos of the newly-typed `Apply` needs to encompass the inlined rhs of the synthetic `ValDef`. `wrappingPos` was made for this.

For right-associative applications with explicit type args, the situation gets pretty hairy in that we have something akin to:
```
argument operator_: [TYPEARG] qualifier
                    ^-Ident-^
         ^-----------Select-----------^
         ^---------TypeApply----------^
^---------------Apply-----------------^
```
that parses to
```
Apply(
  TypeApply(
    Select(qualifier, operator_:),
    Ident(TYPEARG) :: Nil
  ),
  argument :: Nil
)
```

The problem is that the type argument falls within the `Select`, so both the `fun` and `targs` of the `TypeApply` contain that position. `makeTransparent` was made for this. `makeTransparentIf` was made for concision.

Fixes scala/bug#10751.

Also more accurate positions for dynamic invocations. The `Literal` is positioned at the dynamic name and the `Apply`/`TypeApply` reaches over that.

Fixes scala/bug#10203, and likewise closes scala/bug#8562.

Tweaks the advice given on an invalid assignment to work with rangepos.

Position the result of `mkApplies` during update rewrite.

With this commit and scala/scala#6335 (scala/bug#10733), the compiler
bootstraps with rangepos and all the partests pass with it on.

hrhino committed:
Handle rangepos of by-name right-associative applications properly.

And, as a bonus, those with infix type args.

The by-name case is a simple tweak to the SIP-34 translation:
the pos of the newly-typed `Apply` needs to encompass the inlined
rhs of the synthetic `ValDef`. `wrappingPos` was made for this.

For right-associative applications with explicit type args, the
situation gets pretty hairy in that we have something akin to:

```
argument operator_: [TYPEARG] qualifier
                    ^-Ident-^
         ^-----------Select-----------^
         ^---------TypeApply----------^
^---------------Apply-----------------^
```

that parses to

```
Apply(
  TypeApply(
    Select(qualifier, operator_:),
    Ident(TYPEARG) :: Nil
  ),
  argument :: Nil
)
```

The problem is that the type argument falls within the `Select`,
so both the `fun` and `targs` of the `TypeApply` contain that
position. `makeTransparent` was made for this. `makeTransparentIf`
was made for concision.

Fixes scala/bug#10751.

hrhino commented on src/reflect/scala/reflect/internal/Positions.scala:
```diff
@@ -74,7 +74,7 @@ trait Positions extends api.Positions { self: SymbolTable =>
       if (tree.pos.isOpaqueRange) {
         val wpos = wrappingPos(tree.pos, children, focus)
-        tree setPos (if (isOverlapping(wpos)) tree.pos.makeTransparent else wpos)
+        tree setPos tree.pos.makeTransparentIf(isOverlapping(wpos))
```
overeager

hrhino committed:
Throw some wrappingPos on the result of dyna.mkInvoke.

The result is more accurate positions for some dynamics,
as can be seen in the diffs for `run/dynamic-*.scala`,
and less crashy positions for others, as can be seen in
the attached new test case.

Fixes scala/bug#10203.

hrhino committed:
Compare positions in `normalTypedApply`'s error advice by focus.

Follow-up to 1df3796485b4c72affa6eb1c185ec94ed1603798.

Makes sure we get the extra advice to which we're entitled, even
when using rangepos.

hrhino commented:
/sync

hrhino commented:
/rebuild

retronym commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -829,11 +829,14 @@ self =>
+          /* if it's right-associative, `targs` are between `op` and `t` so make the pos transparent */
+          atPos((pos ^| targs.last.pos) makeTransparentIf !treeInfo.isLeftAssoc(op)) {
+            TypeApply(sel, targs)
+          }
```
For my own reference in review:

```scala
  def |(that: Position): Position                   = this union that
  def ^(point: Int): Position                       = this withPoint point
  def |^(that: Position): Position                  = (this | that) ^ that.point
  def ^|(that: Position): Position                  = (this | that) ^ this.point
```

hrhino replied:
I don't "have it in for punctuation", so these were pleasantly expressive to me. If you (of all people!) need to have a cross-reference handy, I'd say that's a sign that they may be less transparent than I had hoped. English names have been restored.

retronym commented on src/reflect/scala/reflect/internal/util/Position.scala:
```diff
@@ -120,6 +120,8 @@ private[util] trait InternalPositionImpl {
   def pointOrElse(alt: Int): Int = if (isDefined) point else alt
   def makeTransparent: Position  = if (isOpaqueRange) Position.transparent(source, start, point, end) else this
+  @inline final def makeTransparentIf(cond: Boolean): Position =
+    if (cond && isOpaqueRange) Position.transparent(source, start, point, end) else this
```
No reason to `@inline` AFAICT.

hrhino replied:
No more `@inline`.

retronym commented on src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:
```diff
@@ -829,11 +829,14 @@ self =>
 
       def mkSelection(t: Tree) = {
-        val pos = opPos union t.pos
+        val pos = (opPos | t.pos) makeTransparentIf !treeInfo.isLeftAssoc(op)
```
perhaps `val rassoc = !treeInfo.isLeftAssoc(op)`, and reuse below.

hrhino replied:
Done.

retronym reviewed (dismissed):
Great work, my comments are only nits, address them if you've got time. The substance of this is already mergeable.

hrhino committed:
move `isLeftAssoc` from `treeInfo` to `nme`, since it's about names, not trees.

hrhino reviewed (commented):
Nits are good to pick. I've updated the PR accordingly.

Also, after thinking on it for a bit:
- I moved the left-associativity detection logic to `nme.isLeftAssoc`, since it's about names, not trees.
- I switched my test to assert things about the actual assigned positions, like the `dynamic-*` tests which I had to change, to increase confidence. (I made it print after `typer` rather than `parser` to show the post-desugaring-and-inlining result; if you'd rather I can also or exclusively dump after `parser`.) There are enough of these that maybe a `PositionTest` in `partest-extras` may be justifiable.

Holler if you think the commits are too granular and I'll squash them. I tried for a logical separation but I also try for a reasonable number of commits.

hrhino commented:
okay, I have gotten the 👍 to do as I will with Scala. Apparently the CLA requires not that they sign it but that I "represent that you have received permission to make Contributions on behalf of that employer", which I have (as of yesterday).

SethTisue commented:
@hrhino welcome back!

hrhino committed:
Position the Apply made for extra implicit args in update rewrite

...  otherwise it's not positioned, that's why ...

With this commit and scala/scala#6335 (scala/bug#10733), the compiler
bootstraps with rangepos and all the partests pass with it on.

retronym committed:
Merge remote-tracking branch 'origin/2.13.x' into review/6423

retronym commented:
I've added a merge from 2.13.x to resolve a merge conflict.

retronym reviewed (approved):
Thanks, keep 'em coming 😄

hrhino commented:
Thanks!
